src/logic/nontrivial.lean

```diff
@@ -128,3 +133,59 @@ begin
   { exact ⟨x₁, (hf.ne_iff' h).2 hx⟩ },
   { exact ⟨x₂, h⟩ }
 end
+
+namespace tactic
+
+/--
+Given a goal `a = b` or `a ≤ b` in a type `α`, generates an additional hypothesis `nontrivial α`
+(as otherwise `α` is a subsingleton and the goal is trivial).
+-/
+meta def nontriviality_by_elim : tactic unit :=
+do
+  t ← (do `(%%a = %%b) ← target >>= whnf, infer_type a) <|> (do `(%%a ≤ %%b) ← target, infer_type a) <|>
```

Comment:
Any reason for `whnf` on `=` but not `≤`?

Reply:
`whnf` will unfold `≤` further (possibly even all the way to type-specific functions)

Reply:
I now use `unify` in all four tests.